### ISSUE TITLE
Guard fused floating-point rewrites with exact VJP and workload gates

### ISSUE THREAD

**Issue by jxsl13** [jxsl13/perfscan#779]
## Finding

Fusing an elementwise loss chain exposed two general analyzer hazards: algebraic equivalence did not preserve raw floating-point behavior, and leaf speedup did not predict workload leverage.

The incumbent EAGLE Smooth-L1 core was:

```text
d = pred - target
core = d*d - ReLU(abs(d)-1)^2
loss = 0.5 * Mean(core)
```

A fused forward needed explicit F32/F64 rounding barriers to match the composed operations for finite extremes, infinities, and NaNs. Go otherwise contracted arithmetic across statements, producing overflow-class and one-bit differences.

The first fused VJP used an algebraically equivalent closed-form derivative. An exact same-binary oracle found one-ULP mismatches in both F32 and F64. Exact parity required reproducing the tape's operation order: form each multiply, add the two fan-out contributions, apply ReLU/Abs selects, then accumulate `(absGrad+p)+p`. Forward and VJP now match the old graph bit-for-bit on special values and 300K parallel tensors.

An order-alternating same-binary benchmark also showed that accelerating only Abs delivered about 1.02x-1.03x at the EAGLE workload, below the 1.03 gate. Fusing the chain produced final forward medians of 1.41x-1.92x and forward+backward medians of 3.02x-3.60x across three 100x/count-7 campaigns.

## Requested perfscan behavior

1. Fusion/reassociation findings should warn that mathematical equivalence is insufficient when replacing an established floating-point graph.
2. Recommend raw-bit forward and VJP oracles with signed zero, infinities, NaNs, finite extremes, and large parallel shapes.
3. Flag same-type conversions or equivalent explicit stores as possible required rounding barriers when contraction changes the contract.
4. Require an order-alternating same-binary workload gate before claiming leverage from an optimized leaf; use an Amdahl shortfall to prioritize the next fusion boundary.

**Comment by jxsl13** [comment]
Implemented by PS6062 in PR #801. That PR was merged into the default branch with all eight CI jobs passing; the current main head is also fully green.

**Comment by jxsl13** [comment]
Related gate-quality finding from GoAI's AMX generation repair: a floating-point tolerance assertion of the form `if math.Abs(got-want) > limit { fail }` silently accepts a NaN result. It also cannot prove that a NaN-poisoned destination was overwritten. This can let an invalid fast kernel qualify as a performance win.

Concrete public example: [`checkTolerance` at the failing PR revision](https://github.com/jxsl13/goai/blob/bdb2f603cd54295a18ff732bcc3c9259c66dca0e/backend/cpu/gemm_amx_test.go#L26-L41). The associated test pre-fills C with NaNs. The generation repair now requires explicit finite-result checks for finite fixtures, explicit NaN/Inf class comparisons for nonfinite fixtures, and signed-zero checks where bit identity is promised, before a tolerance comparison.

Suggested extension to this issue: performance-validation guidance should flag a tolerance-only rejection predicate lacking dominating finite/class checks, particularly when output buffers are poisoned or the candidate uses native kernels. A detector should report an advisory with both the assertion and initialization sites, avoid assuming every tolerance helper is unsafe when callers already constrain values, and never auto-rewrite the numeric policy. Validate with a reached NaN-output mutation that must fail. This is a validation-quality opportunity, not a speedup claim.
